### DISCUSSION TITLE
Retrieving nuget.config files using MSBuild

### DISCUSSION THREAD

**Discussion by wuhkuh** [NuGet/Home#11464] [Q&A] [answered]
Hey!

I would like to get all `nuget.config` files used for a restore operation in an MSBuild task.
Currently, the approach I'm looking at is parsing the output of `RestoreTaskEx`, but I feel there might be a better solution.
Is there a better solution?

Thanks!

**Comment by zivkan** [comment, marked as answer]
If your project uses PackageReference, then in your obj folder, you'll have a `project.assets.json` file which will list all the config files in there somewhere. Note that since it's in the obj (intermediate) folder, we don't promise backwards compatibility, we might change the format as needed, although it's very unlikely that we'll break the format, and never have so far (but we have added new information in the past, making old versions of Nuget restore incompatible with new versions of msbuild).  If your projects are using packages.config, there's no equivalent. You could add an empty project, that uses PackageReference, just for the project.assets.json file, if getting the list of config files is sufficiently important to you.

Another option is to use the NuGet.Configuration package from nuget.org, and `Settings.LoadDefaultSettings(string path)` will load all nuget.config fies in the same way that restore does. Pass the directory containing the solution file as the parameter. Then you'll have a setting object. I don't know how to get the list of all nuget.config files from that object, but there must be a way, since restore will list the files. Note, in the last year or two, we did add a new location where nuget searches for nuget.config files, so if you're doing this for any kind of auditing reasons, you'll need to make sure your version of NuGet.Configuration.dll keeps up to date with whatever build tools you're using, in case we add more in the future. Very unlikely, but auditing tools often don't tolerate missing/inconsistent data.

Those are the only two suggestions I can think of right now.

**Reply by wuhkuh** [reply]
Thanks! That got me up to speed 😄
I went with the second option, the only difference being dynamically loading `NuGet.Configuration.dll` provided with the SDK, to prevent compatibility issues 🙂
